### ISSUE TITLE
package-updates-lib.pl doesn't look at snaps

### ISSUE THREAD

**Issue by superian** [webmin/webmin#1864]
I know many of us really do not like snaps, but Ubuntu users are more or less stuck with them.

Currently, as Ubuntu is a Debian-based system, the system packages bits of webmin only use `apt` to check for / install updates.

The `snap` command does enable users to see if any updates to installed snaps are available and, if asked, to install them.

**Comment by jcameron** [comment]
Is the `snap` command a replacement for `apt` ? 

Also which Ubuntu version are you running there?

**Comment by swelljoe** [comment]
Snap is not a replacement for apt/dpkg. It is a container-based package format, similar to [Flatpak](https://www.flatpak.org/). [About snap.](https://snapcraft.io/about) Both are somewhat comparable to Docker, though probably better suited to use as a "package" format (and both are more open than Docker, which has become even more terrible recently).

I've ruled out using snaps for anything we do with Virtualmin/Cloudmin, as there is no Open Source way to host snap repositories. Flatpak can have an entirely Open Source tool-chain, from building to publishing. We'd have to pay Canonical to host our software repositories, which is simply not feasible (and, also toxic, we shouldn't have to pay to make and distribute software that makes their OS more valuable).

Snap seems primarily like a mechanism for Canonical to generate revenue, frankly. But, OP is right that they have begun moving in the direction of doing more with snaps; several packages in the Ubuntu standard apt repos are now snaps wrapped in a deb package (but in those cases, you update with apt, so no additional support is needed). Chromium is one such package. But, most snap packages use snap to manage them, instead of wrapping them in a deb and putting them in an apt repo.

I'm not opposed to adding snap awareness, I guess, but I also personally discourage their use, on principle. If we're adding snap support, I'd suggest adding flatpak first, since it is a more open option for container-based application deployment, and they're similar enough that by the time one is implemented, the other should be easy.

**Comment by jcameron** [comment]
Ah ok ... so it sounds like support for Snap and similar container systems would necessitate a whole new module, not just a new mode in the Software Packages module.

**Comment by iliajie** [comment]
> Currently, as Ubuntu is a Debian-based system, the system packages bits of webmin only use apt to check for / install updates.

`snap` updates supposed to be happening automatically (every 4 hours by default). You can check this by running `snap refresh --time` command.

> Ah ok ... so it sounds like support for Snap and similar container systems would necessitate a whole new module, not just a new mode in the Software Packages module.

Why not. `snap` has a list of pretty standard commands to list, install and update software. It shouldn't be any much different from using `apt` command, right?

The one thing I'm not certain about if `snap` can/should be a complete replacement for `apt` or they expected to work together. I personally used `snap` alongside with a standard package manager (`dnf` in Fedora).

**Comment by swelljoe** [comment]
I don't think it's different enough to warrant its own module. It _is_ a package manager, it just deploys them via containers.

**Comment by swelljoe** [comment]
> The one thing I'm not certain about if snap can/should be a complete replacement for apt or they expected to work together.

It is not a replacement for `apt`, at least not for the foreseeable future.

>  I personally used snap alongside with a standard package manager (dnf in Fedora).

That doesn't even make sense! Flatpak is already there in Fedora, and technically superior (plus a wholly Open Source toolchain).

**Comment by iliajie** [comment]
> That doesn't even make sense! Flatpak is already there in Fedora, and technically superior (plus a wholly Open Source toolchain).

Perhaps, but I couldn't figure out in timely manner how to install the latest version of `hugo` back in time ..

**Comment by superian** [comment]
As I say, snaps are not everyone's preferred choice of package manager, but Canonical are more or less imposing them on Ubuntu users, where they are an **addition** to the standard Debian apt family of package management programs. Amongst other things, the software that livepatches the kernel on Ubuntu has been provided as a snap since Ubuntu 20.04, I think, and is installed by default on a new Ubuntu system with, I think, no other way to get it.

I don't think there is a need for another module to cope with them as the command line interface is simple: `snap refresh` as root will update any and all installed snaps that have updates, for example.

If I have asked webmin to install updates automatically, for example, it would be nice if that were added to the process when they are done.

(I don't think it would be possible, never mind wise, for snaps to replace apt entirely. I am under the impression that Flatpak is going to be uncommon on servers, as I believe that it's for GUI programs only where as snap also works for command line programs.)

**Comment by superian** [comment]
> `snap` updates supposed to be happening automatically (every 4 hours by default). You can check this by running `snap refresh --time` command.

That's right, however - certainly for some desktop apps - Canonical have taken the decision that it's ok to manage their resourcing of the snap store servers by delaying at least some users getting available updates automatically for up to a couple of days unless they ask for them by running `snap refresh`.

It's another reason why some of us do not like snaps, but if you want / have to have a Ubuntu server, you are more or less stuck with having them as part of the whole package.

**Comment by iliajie** [comment]
> I don't think it would be possible, never mind wise, for snaps to replace apt entirely.

In this case then I think we shouldn't add it to the Package Updates module, and it should be technically a separate module because `snap` can be used on any Linux distro standalone.

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
